csrc/ir/nodes.cpp

```diff
@@ -389,9 +389,81 @@ UnaryOp::UnaryOp(IrBuilderPasskey passkey, UnaryOpType type, Val* out, Val* in)
 
 std::vector<PolymorphicValue> UnaryOp::evaluate(
     const ExpressionEvaluator& ee,
-    const std::vector<PolymorphicValue>& inputs) const {
+    std::unordered_map<const Val*, PolymorphicValue>& known_values) const {
   using namespace PolymorphicValue_functions;
-  const auto& in = inputs.at(0);
+
+  // If the UnaryOp is CastOp, check if the preceding pattern of
+  // operators matches with matmul (MmaOp(Broadcast (A), Broadcast(B)) -> Cast)
+  // or matmul + bias (BinaryOp::Add (MmaOp(Broadcast (A), Broadcast(B),
+  // Broadcast(bias)) -> Cast) If not, evaluate UnaryOp::CastOp along with the
+  // other types by evaluating the immediate input.
+
+  auto has_matmul_and_bias = [](Val* in) -> bool {
+    if (auto* binary = dynamic_cast<BinaryOp*>(in->definition())) {
+      if (binary->getBinaryOpType() == BinaryOpType::Add) {
+        if (binary->input(0)->definition()->isA<MmaOp>()) {
+          return true;
+        }
+      }
+    }
+    return false;
+  };
+
+  if ((getUnaryOpType() == UnaryOpType::Cast) &&
+      input(0)->definition() != nullptr) {
+    // Case 1: MmaOp + Cast (Matmul)
+    if (auto* mma = dynamic_cast<MmaOp*>(input(0)->definition())) {
+      MmaOpUtils::verifyMmaOpForEvaluation(mma, out()->getDataType().value());
+
+      std::vector<at::Tensor> mma_inputs;
+      for (Val* inp : mma->inputs()) {
+        auto eval_i = ee.evaluate(inp, known_values);
+        mma_inputs.push_back(eval_i.as<at::Tensor>());
+      }
+
+      const auto a = mma_inputs.at(0).squeeze(-1);
+      const auto b = mma_inputs.at(1).squeeze(0);
+
+      // After removing the broadcast dimensions, the format should be
+      // [M, K] x [K, N] compatible with aten::matmul format.
+      return {a.matmul(b)};
+    }
+
+    // Case 2: Matmul + Bias
+    if (has_matmul_and_bias(input(0))) {
```

Comment:
In this version, the cast+broadcast+add+matmul+cast pattern we are looking for is checked in has_matmul_and_bias and is assumed outside (line 434 and 435). This creates room for error in the future when someone changes one place but forgets to change the other. I think it's going to be cleaner and more robust to combine matching the pattern and assigning the matched results into one function. This is remotely analogous to `re.match`ing a pattern with groups -- `re.match` both matches the pattern and associates the matched results with the group ID. 

```
Val* mma_lhs = nullptr;
Val* mma_rhs = nullptr;
if (matchMatmulCast(this, std::ref(mma_lhs), std::ref(mma_rhs))) {
  ...
  return ...;
}

Val* bias = nullptr;
if (matchMatmulBiasCast(this, std::ref(mma_lhs), std::ref(mma_rhs), std::ref(bias))) {
  ...
  return ...;
}

Evaluate the immediate inputs and this op. 
```

Yes, the root CastOp will be checked twice, but at this moment I'd prefer simplicity to performance unless we prove that the extra check is significant overhead (I don't think it is).

Reply:
It may simplify to merge (some of) `verify*` into `match*` as well.

Reply:
I am somewhat unclear on your concern. We do check that the bias side satisfies all the required conditions in `verifyBiasForEvaluation` similar to how the required assumptions for MmaOp are enforced in `verifyMmaForEvaluation.`

Are you suggesting merging the `has_matmul_and_bias` with assignment of bias? Or the verification should be done here as well instead of another call?

Reply:
> Are you suggesting merging the has_matmul_and_bias with assignment of bias? Or the verification should be done here as well instead of another call?

Both. Let me clarify a bit more: I'd love to see everything about that pattern (including matching the pattern, retrieving the matched results, and probably verification about that pattern) to be localized, so it's harder to make oops-I-forgot-to-also-change-this type of programming errors and easier to read the code without jumping in and out of the `has_matmul_and_bias` function. 

I'm happy to chat more about it after 2:30 today!

Reply:
Okay, I think I get your point. I still prefer keeping the verification of `MmaOp` modular since that will be reused in all patterns.
Let me refactor this, and we can discuss more.

Reply:
> the verification of MmaOp modular since that will be reused in all patterns.

That makes sense. I guess we can call that function from both matchMatmulCast and matchMatmulBiasCast. Anyhow, I'm looking forward to the next version so the discussion is more concrete!